File: MathCounts/2020/HankSequence.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stdio.h>

using namespace std;

bool isNext(int n, int M, vector<int>& result)
{
    for (size_t i = result.size()-1; i > 0; --i)
    {
        int diff = n - result[i];
        for (size_t j = 0; j < i; ++j)
        {
            if (result[i] - result[j] == diff)
            {
                if (M == n)
                {
                     std::cout << result[j]  << ", " << result[i] << ", and " << n << " form an arithmetic sequence." << std::endl;
                }
                return false;
            }
        }
    }
    
    return true;
}

std::string numberInBase(int n, int b)
{
    std::string result;
    if (n == 0) return "0";

    int count = 0;
    while (count < 6)
    {
      if (n == 0) {
         return result;
      }

      int rem = n%b;
      result = std::to_string(rem) + result;
      n = n/b; 
      count++;
    }

    return result;
}


int main(int argc, char* argv[])
{
    vector<int> result;
    result.push_back(1);
    result.push_back(2);
    int N = 729;
    int Base = 10;
    bool subtract = false;
    int M = 0;

    for (int i = 1; i < argc; ++i)
    {
        if (std::string(argv[i]) == "-upper" && i+1 < argc)
        {
            N = std::stoi(argv[i+1]);
        }
        else if (std::string(argv[i]) == "-base" && i < argc -1)
        {
            Base = std::stoi(argv[i+1]);
        }
        else if (std::string(argv[i]) == "-subtract")
        {
            subtract = true;
        }
        else if (std::string(argv[i]) == "-check" && i+1 < argc)
        {
            M = std::stoi(argv[i+1]);
        }
    }

    for (int i = 3; i <= N; ++i)
    {
        bool canAdd = isNext(i, M, result);
        if (canAdd) result.push_back(i);
    }

    if (M > 0) return 0;

    std::cout << "Number of terms: " << result.size() << std::endl;

    for (size_t i = 0; i < result.size(); ++i)
    {
        if (i != 0 && i%8 == 0) std::cout << std::endl;

        int n = subtract ? result[i] - 1 : result[i];
        std::string v = numberInBase(n, Base);
        if (Base == 10) 
        {
            std::string w("      ");
            w = w.substr(0, 6-v.length()) + v;
            std::cout << w;
        }
        else
        {
            std::string w("000000");
            w = w.substr(0, 6-v.length()) + v;
            std::cout << w << " ";
        }
    }

    std::cout << std::endl;
    return 0;
}

File: MathCounts/2020/HankSequence.java
import java.util.*;

class HankSequence
{
    public static void main(String[] args) {

        int base = 10;
        int n = 729;
        int M = 0;
        boolean subtract = false;
        int start = 1;

        if (args != null && args.length > 0) {
            for (int i = 0; i < args.length; ++i)
            {
                if (args[i].compareTo("-upper") == 0 && i < args.length - 1) {
                    n = Integer.parseInt(args[i+1]);
                }
                else if (args[i].compareTo("-base") == 0 && i < args.length - 1)
                {
                    base = Integer.parseInt(args[i+1]);
                }
                else if (args[i].compareTo("-check") == 0 && i < args.length -1)
                {
                    M = Integer.parseInt(args[i+1]);
                }
                else if (args[i].compareTo("-subtract") == 0) {
                    subtract = true;
                }
                else if (args[i].compareTo("-start") == 0 && i < args.length - 1)
                {
                    start = Integer.parseInt(args[i+1]);
                }
            }
        }

        HankSequence h = new HankSequence();

        ArrayList<Integer> result = h.BuildHankSequence(start, n, M);

        if (M > 0) return;

        System.out.println("There are " + result.size() + " terms:");

        for (int i = 0; i < result.size(); i++)
        {
            if (i % 8 == 0 && i != 0) System.out.println();
            int x = subtract ? result.get(i) - 1 : result.get(i);
            String numberAsString = Integer.toString(x, base);
            String s = base == 10? "        ".substring(numberAsString.length()) + numberAsString : "000000".substring(numberAsString.length()) + numberAsString;
            System.out.print(s + " ");
        }
    }

    ArrayList<Integer> BuildHankSequence(int start, int n, int M)
    {
        ArrayList<Integer> result = new ArrayList();
        result.add(start);
        result.add(start+1);

        for (int i = start + 2; i < n; i++)
        {
            boolean canAdd = IsValidNextNumber(i, M, result);
            if (canAdd) {
                result.add(i);
            }
        }

        return result;
    }

    boolean IsValidNextNumber(int n, int M, ArrayList<Integer> result)
    {
        for (int i = result.size()-1; i >= 0; i--)
        {
            int diff = n - result.get(i);
            for (int j=0; j < i; j++)
            {
                if (result.get(i) - result.get(j) == diff)
                {
                    if (M == n)
                    {
                        System.out.print(result.get(j) + ", " + result.get(i) + ", and " + n + " form an arithmetic sequence.");
                    }
                    
                    return false;
                }
            }
        }
    
    return true;
    }
}